### ISSUE TITLE
reproducing results ...

### ISSUE THREAD

**Issue by ppisljar** [MShahabSepehri/CryptoMamba#8]
Hello,

first thank you for your work! 

i was trying to reproduce the results from this repository and I found some unexpected behaviour which i hope you can help to clarify.

1. reported trade gains dont take into the account actual bitcoin price movement. For example for your test split bitcoin went up more than 100%, and making a single trade (buying at the beggining of test period and selling at the end) will get you a big gain already.  To show how much better this model performs it would be nice to show the actual difference in performance against this base line.

```
cmamba_nv -- Total gain in %: 122.52
cmamba_nv -- Price increase in timerange: 28010.0 - 59993.03125 = 31983.03
cmamba_nv -- Price increase in %: 114.18
cmamba_nv -- Trading gain diff in %: 8.33

```

2. predictions seem to be "fixed" somehow as i can observe this:
- running the same model multiple times over the same time range produces EXACTLY the same results. (i would expect some variance)
- training a new model on the same dataset produces EXACTLY the same results (i would expect even more variance)
- training a new model on a new dataset (same source, but data from 2014-2023 was used for train set) produces EXACTLY the same results (here i would definetely expect very different results)

```
$ python scripts/simulate_trade.py --config cmamba_nv --ckpt_path checkpoints/cmamba_nv.ckpt 
350 data points loaded as test split.
['Open', 'High', 'Low', 'Close', 'Volume', 'Timestamp']
['Open', 'High', 'Low', 'Close', 'Volume', 'Timestamp']
['Open', 'High', 'Low', 'Close', 'Volume', 'Timestamp']
['Open', 'High', 'Low', 'Close', 'Volume', 'Timestamp']
cmamba_nv -- Final balance: 222.52
cmamba_nv -- Maximum Draw Down : 26.15
$ python scripts/simulate_trade.py --config cmamba_nv --ckpt_path logs/CMamba/version_6/checkpoints/epoch1006-val-rmse576.8903.ckpt 
350 data points loaded as test split.
['Open', 'High', 'Low', 'Close', 'Volume', 'Timestamp']
['Open', 'High', 'Low', 'Close', 'Volume', 'Timestamp']
['Open', 'High', 'Low', 'Close', 'Volume', 'Timestamp']
['Open', 'High', 'Low', 'Close', 'Volume', 'Timestamp']
cmamba_nv -- Final balance: 222.52
cmamba_nv -- Maximum Draw Down : 26.15
$ python scripts/simulate_trade.py --config cmamba_nv --ckpt_path logs/CMamba/version_7/checkpoints/epoch1079-val-rmse1518.6656.ckpt 
350 data points loaded as test split.
['Open', 'High', 'Low', 'Close', 'Volume', 'Timestamp']
['Open', 'High', 'Low', 'Close', 'Volume', 'Timestamp']
['Open', 'High', 'Low', 'Close', 'Volume', 'Timestamp']
['Open', 'High', 'Low', 'Close', 'Volume', 'Timestamp']
cmamba_nv -- Final balance: 222.52
cmamba_nv -- Maximum Draw Down : 26.15
```

3. running the model on any other time period produces very different (not so positive) results
- running any of the localy trained models on this other time frame again produces the same results

```
$ python scripts/simulate_trade.py --config cmamba_nv6 --ckpt_path logs/CMamba/version_7/checkpoints/epoch1079-val-rmse1518.6656.ckpt 
91 data points loaded as test split.
['Open', 'High', 'Low', 'Close', 'Volume', 'Timestamp']
['Open', 'High', 'Low', 'Close', 'Volume', 'Timestamp']
['Open', 'High', 'Low', 'Close', 'Volume', 'Timestamp']
cmamba_nv6 -- Final balance: 146.29
cmamba_nv6 -- Maximum Draw Down : 11.88
$ python scripts/simulate_trade.py --config cmamba_nv6 --ckpt_path logs/CMamba/version_6/checkpoints/epoch1006-val-rmse576.8903.ckpt 
91 data points loaded as test split.
['Open', 'High', 'Low', 'Close', 'Volume', 'Timestamp']
['Open', 'High', 'Low', 'Close', 'Volume', 'Timestamp']
['Open', 'High', 'Low', 'Close', 'Volume', 'Timestamp']
cmamba_nv6 -- Final balance: 146.29
cmamba_nv6 -- Maximum Draw Down : 11.88
$ python scripts/simulate_trade.py --config cmamba_nv6 --ckpt_path checkpoints/cmamba_nv.ckpt 
91 data points loaded as test split.
['Open', 'High', 'Low', 'Close', 'Volume', 'Timestamp']
['Open', 'High', 'Low', 'Close', 'Volume', 'Timestamp']
['Open', 'High', 'Low', 'Close', 'Volume', 'Timestamp']
cmamba_nv6 -- Final balance: 146.29
cmamba_nv6 -- Maximum Draw Down : 11.88

```

**Comment by ppisljar** [comment]
this are my configs:
`cmamba_nv`: original config which was used to train `checkpoints/cmamba_nv.ckpt`
`cmamba_nv5`: original config, was used to train `version5` which should be similar to published model (all params and data are the same)
`cmamba_nv6`: changes the data config to use this: (data is fetched using yfinance for larger time period)

```
jumps: 1440
start_date: "2014-17-09"
end_date: "2025-24-02"
root: "./data"
data_path: "./data/BTC_USDT_day.csv"
train_interval: ["2014-17-09", "2023-17-09"]
val_interval: ["2023-17-09", "2024-17-09"]
test_interval: ["2024-17-09", "2025-24-02"]
```

**Comment by AsalMehradfar** [comment]
Hi,

Thank you for your feedback and for reproducing our results! Below are responses to your concerns:

---

### 1. **Buy-and-Hold vs. Model Performance**  
   - You are absolutely right that the buy-and-hold strategy during our test period results in a significant gain due to Bitcoin's price increase.  
   - To address this concern, we have recently added a **long-short trading strategy** (allowing for negative BTC positions), which better demonstrates our model's effectiveness beyond simple price appreciation.  
   - Additionally, in financial trading, risk management is as important as returns. To provide a more comprehensive evaluation, we have included the **Maximum Drawdown (MMD)** metric, showing that CryptoMamba effectively reduces MMD from **26% to 13%**, mitigating downside risk.  
   - If you examine different test periods where Bitcoin’s final price is not substantially higher, you will see a more pronounced advantage of our model compared to the buy-and-hold strategy.

---

### 2. **Model Predictions Appearing Fixed for Multiple Running or Training on The Same Data**  
   - A trained model is a **deterministic method**, meaning that running it with the same inputs will always yield the same outputs.  
   - For **reproducibility**, we explicitly set a **random seed** at the beginning of the training code. If the same seed is used, the model will be trained in exactly the same way, producing identical results.  You can change the seed yourself in different runs if you want to train different models.
 
---

### 3. **Possible Misconfiguration in `simulate_trade.py`**  
   - It seems that you may have **changed the default argument** of `trade_mode` in `simulate_trade.py`, which is `smart` by default. The results you are seeing correspond to a **buy-and-hold strategy** (`no_strategy`), which explains why changing the model does not affect the output.  
   - Additionally, we noticed that you are **training/loading an `nv` model** (which is designed for the **no-volume case**) while simultaneously **loading data that includes volume**. This mismatch suggests that you are using the wrong model with the wrong dataset. This should have triggered an error during training.  
   - To ensure correct results, please verify that you are using the appropriate model configuration and dataset. When cloning the repository and running the code without any modifications, the expected output should be:  

     ```bash
     $ python scripts/simulate_trade.py --config cmamba_nv --ckpt_path checkpoints/cmamba_nv.ckpt
     ['Timestamp', 'Open', 'High', 'Low', 'Close']
     350 data points loaded as test split.
     cmamba_nv -- Final balance : 184.56
     cmamba_nv -- Maximum Draw Down : 16.01
     ```

### 4. **Unexpected Results on Different Time Frames**  
- We still believe that the results you are seeing correspond to **buy-and-hold** rather than active trading results. We strongly suggest first fixing the `trade_mode` argument or cloning the repository from scratch to eliminate any potential configuration errors.  
- Additionally, we recommend using **models with volume**, as they have demonstrated **higher accuracy** in our paper.  
- For reference, the **published models** are:
  - **v1**: Without Volume  
  - **v2**: With Volume  
  - The rest are **prototypes** and may not be fully validated.  

---

Let us know if you have any further questions!

**Comment by ttys0001** [comment]
I have similar questions for this code.  I use the following code to train and print the test result is differnet from the evaluation result. Could you know the reason? Thank you very much.

python3 scripts/training.py --config cmamba_v
python scripts/evaluation.py --config cmamba_v --ckpt_path logs/CMamba/version_0/checkpoints/epoch949-val-rmse571.4828.ckpt

For example, here is the result in the training log.
CMamba-v
test/mae            1124.4158935546875
test/mape          0.020408714190125465
test/mse                2571888.25
test/rmse            1567.802001953125

SMamba-v
test/mae             1207.349853515625
test/mape          0.022085757926106453
test/mse                 2718316.5
test/rmse           1610.5946044921875

GRU-v
test/mae            0.02079002372920513
test/mape           0.02683136612176895
test/mse           0.0008007054566405714
test/rmse          0.027869226410984993

But the file of evaluation.py outputs is different. Here is the result.

CMamba-v
 Split        MSE          RMSE     MAPE      MAE    
 Train    1383240.625    1176.113  0.02754   670.88  
  Val     359435.875      599.53   0.01674  396.453  
 Test     2566110.25     1601.908  0.02039  1122.959 

SMamba-v
 Split        MSE          RMSE     MAPE      MAE    
 Train     1117376.5     1057.06   0.0285   631.626  
  Val     445444.312     667.416   0.01969   465.05  
 Test     2714342.75     1647.526  0.02208  1206.898 

GRU-v
 Split        MSE          RMSE     MAPE      MAE    
 Train     1267630.0     1125.891  0.03112  659.947  
  Val     338008.969     581.385   0.01569  371.517  
 Test      3894030.0     1973.33   0.02519  1449.952

**Comment by nico7091** [comment]
Hello,

first thank you for your work!

I have reviewed your model and am working on it. The question is, with all that has happened and all that has been done, do you think it makes sense for us to spend a lot of time and money on this idea or to use the usual methods with transformers and GRU for example?
@AsalMehradfar 
Because when I consulted with a few people, they said that the Model MAMBA is currently too basic and too complicated.

Thank you in-advance

**Comment by AsalMehradfar** [comment]
Hi @ppisljar ,

Thank you for your pull request!

Since our code is based on the **CryptoMamba** paper, which includes a dedicated study on the impact of **volume**, we decided to keep volume with a separate flag rather than merging it with the additional features. This makes it easier to reproduce and compare results consistently with our experiments.

To accommodate your suggestion, we’ve updated the codebase to support a new `additional_features` parameter. You can now specify any extra features you want to use (e.g., fear and greed index) by adding a list under `additional_features` in both the **data** and **training** configuration files.

Meanwhile, if you want to include **volume**, you should continue setting `use_volume = True`.

We’ve committed these changes separately and have therefore **closed the pull request**, but we truly appreciate your contribution and the suggestion—it was a helpful idea!

If you plan to use features beyond the standard ones (`Open`, `Close`, `High`, `Low`, `Timestamp`, and `Volume`), just make sure to:
- Include them in your **raw dataset**, and
- List them under `additional_features` in the config files.

Thanks again for contributing!

**Comment by AsalMehradfar** [comment]
Hi @ttys0001 , 

Thank you for your question!

There are two main reasons why the loss values from `training.py` and `evaluation.py` appear different:

### 1. Difference in How Loss is Computed
During training, the loss is computed using **PyTorch Lightning’s default logging mechanism**, which often averages the loss **across batches**, not across the entire dataset. In contrast, our `evaluation.py` script uses **our own custom evaluation logic**, which computes the metrics on the **entire dataset** after denormalizing the outputs. This leads to slight differences in reported metrics.

### 2. Previously Normalized Loss Output
You also noticed a larger discrepancy specifically for the **GRU** model. That’s because in earlier versions of the code, for models like GRU that were trained with normalized values, the reported metrics were also computed on the normalized scale. This made GRU’s losses appear much smaller than they actually were as they were scaled by a constant factor.

To avoid any confusion and maintain consistency across all models, we’ve updated the code to ensure that **all reported losses are based on unnormalized (real-world) values**.

We recommend pulling the latest version of the code for the updated behavior.

---

Since this message is under someone else's issue thread and your question is now addressed, I’ll go ahead and **close this thread**. If you have any other questions, feel free to **open a new issue** anytime.

Thanks again!
